modules/console/pages/ui/programmable-push-filters.adoc

```diff
@@ -2,58 +2,151 @@
-{ui} can inject the following properties into your function, which you can use in your filter code:
+== Add a JavaScript filter
 
-* `partitionId` - The record's partition ID
-* `offset` - The record's offset within its partition
-* `key` - The record's key in its decoded form
-* `value` - The record's value in its decoded form
-* `headers` - The record's header value in its decoded form
+To add a JavaScript filter:
 
-NOTE: Keys, values, and headers are passed into your JavaScript code in their decoded form. The
-deserialization logic (for example, decode an Avro serialized byte array to a JSON object) is applied first, before injecting it into
-the JavaScript function. If your record is presented as a JSON object in the UI, you can also access it
-like a JavaScript object in your filter code.
+. Navigate to the topic's *Messages* page.
+. Click *Add filter > JavaScript Filter*.
+. Define your JavaScript filtering logic in the provided input area.
 
-Suppose you have a series of Avro, JSON, or Protobuf encoded record values that deserialize to JSON objects like this:
+image::ROOT:console:js-filter.png[alt="JavaScript filter in {ui}"]
 
-[,json]
+== Resource usage and performance
+JavaScript filters are executed on the backend, consuming CPU and network resources. The performance of your filter depends on the complexity of your JavaScript code and the volume of data being processed.
+Typical performance is *15,000-20,000 filtered records/second per CPU core*.
+Complex JavaScript logic or large data volumes may increase CPU load and network usage.
+
+== Available JavaScript properties
+
+{ui} injects these properties into your JavaScript context:
+
+[cols="1a,2a,1a"]
+|===
+| Property         | Description                                      | Type
+
+| `headers`        | Record headers as key-value pairs (ArrayBuffers) | Object
+| `key`            | Decoded record key                               | String
+| `keySchemaID`    | Schema Registry ID for key (if present)          | Number
+| `partitionId`    | Partition ID of the record                       | Number
+| `offset`         | Record offset within partition                   | Number
+| `timestamp`      | Timestamp as JavaScript Date object              | Date
+| `value`          | Decoded record value                             | Object/String
+| `valueSchemaID`  | Schema Registry ID for value (if present)        | Number
+|===
+
+NOTE: Values, keys, and headers are deserialized before being injected into your script.
+
+== JavaScript filter examples
+
+=== Filter by header value
+
+*Scenario:* Records tagged with headers specifying customer plan type.
+
+.Sample header data
+[source,json]
+----
+headers: {
+  "plan_type": "premium"
+}
+----
+
+.JavaScript filter
+[source,javascript]
+----
+let headerValue = headers["plan_type"];
+if (headerValue) {
+  let stringValue = String.fromCharCode(...new Uint8Array(headerValue));
+  return stringValue === "premium";
+}
+return false;
+----
+
```

Comment:
Maybe another example where a header value is JSON...
```json
headers: {
  "customer": "{\"orgID\":\"123-abc\",\"name\":\"ACME Inc.\"}"
}
```

```js
let headerValue = headers["header0"]
// filter if value in header is object with key "foo" and the value is "bar"
if (headerValue != null && headerValue != undefined) {
    let stringValue = String.fromCharCode(...headerValue);
    ley valueObj = JSON.Parse(stringValue)
    return valueObj["orgID"] == "123-abc"
}

return false
```

Reply:
Added 👍